File: programs/index/client/createIndex.ts
import { Program } from "@project-serum/anchor";
import {
  PublicKey,
  SystemProgram,
  TransactionInstruction,
} from "@solana/web3.js";
import { IndexProgram } from "../../../target/types/index_program";
import { PDA } from "../../../utils";

export type CreateIndexProps = {
  indexPDA: PDA;
  // signer: PublicKey;
  owner: PublicKey;
  namespace: String;
  isSerial: boolean;
};

export function createIndex(
  indexProgram: Program<IndexProgram>,
  { indexPDA, owner, namespace, isSerial }: CreateIndexProps
): // { indexPDA, signer, owner, namespace, isSerial }: CreateIndexProps
TransactionInstruction {
  return indexProgram.instruction.createIndex(
    // owner,
    namespace,
    isSerial,
    indexPDA.bump,
    {
      accounts: {
        index: indexPDA.address,
        payer: owner,
        owner: owner,
        systemProgram: SystemProgram.programId,
      },
    }
  );
}

File: programs/index/src/instructions/create_index.rs
use {
    anchor_lang::{
        prelude::*,
        solana_program::system_program
    },
    crate::state::*,
    std::mem::size_of
};

#[derive(Accounts)]
#[instruction(
    owner: Pubkey,
    namespace: String,
    is_serial: bool,
    bump: u8,
)]
pub struct CreateIndex<'info> {
    #[account(
        init, 
        seeds = [
            SEED_INDEX, 
            owner.as_ref(),
            namespace.as_ref()
        ],
        bump = bump, 
        payer = signer, 
        space = 8 + size_of::<Index>()
    )]
    pub index: Account<'info, Index>,

    #[account(mut)]
    pub signer: Signer<'info>,
    
    #[account(address = system_program::ID)]
    pub system_program: Program<'info, System>,
}

pub fn handler(
    ctx: Context<CreateIndex>, 
    owner: Pubkey,
    namespace: String,
    is_serial: bool,
    bump: u8,
) -> ProgramResult {
    // Get accounts.
    let index = &mut ctx.accounts.index;

    // Initialize index account.
    index.owner = owner;
    index.namespace = namespace;
    index.count = 0;
    index.is_serial = is_serial;
    index.bump = bump;
    
    return Ok(());
}